keps/sig-scheduling/4816-dra-prioritized-list/README.md

```diff
@@ -0,0 +1,1130 @@
+`deviceClassName`, because it has no alternative request types. The "gpu"
+request could be met by several different types of GPU, in the listed order of
+preference. Each of those is a separate `DeviceRequest`, with both a
+`deviceClassName` and also its own name. The fact that these subrequests also
+have their own names allows us to apply constraints or configuration to
+specific, individual subrequests, in the event that it is the chosen
+alternative. In this example, the "small-gpu" choice requires a configuration
+option that the other two choices do not need. Thus, if the resolution of the
+"gpu" request is made using the "small-gpu" subrequest, then that configuration
+will be attached to the allocation. Otherwise, it will not.
+
+Similarly, for `Constraints`, the list of requests can include the main request
+name ("gpu" in this case), in which case the constraint applies regardless of
+which alternative is chosen. Or, it can include the subrequest name, in which
+case that constraint only applies if that particular subrequest is chosen.
+
+In the PodSpec, however, the subrequest names are not valid. Only the main
+request name may be used.
+
+### Scheduler Implementation
+
+Currently, the scheduler loops through each entry in `DeviceClaim.Requests` and
+tries to satisfy each one. This would work essentially the same, except that
+today, it [throws an
+error](https://github.com/kubernetes/kubernetes/blob/03f134461462f86239067ec20ec17a0ba892db52/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go#L164)
+when it encounters a claim with a missing `DeviceClassName`. Instead, here we
+would check for entries in `FirstAvailableOf`, and add an additional loop,
+trying each of these requests in order.
+
+The current implementation will navigate a depth-first search of the devices,
+trying to satisfy all requests and contraints of all claims. The optionality
+offered at the `DeviceRequest` level provides another index state to track in
+the
+[`requestIndices`](https://github.com/kubernetes/kubernetes/blob/03f134461462f86239067ec20ec17a0ba892db52/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go#L362) and [`deviceIndices`](https://github.com/kubernetes/kubernetes/blob/03f134461462f86239067ec20ec17a0ba892db52/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go#L368). In the case of the feature gate
+disabled, this new index will always be 0.
+
+Alternatively, we can refactor to make this code more defensible via a feature
+gate.
+
+DRA today works on a "first match" basis for a given node. That would not change
+with this KEP. However, in order for the scheduler to prefer a node that has the
+initial prioritized device request, those requests would need a higher score.
+This will be implemented for beta. For alpha, the scheduler may still pick a
+node with a less preferred device, if there are nodes with each type of device
+available.
```

Comment:
Can you clarify this a bit. The way it reads it seems like you are saying that today the list isn't *actually* prioritized. Meaning the scheduler is free to pick any of the possible choices, even if a device higher in the list is available.

Reply:
I will. But given how the scheduler works (evaluate fit and score each node), and the fact that nodes do not usually have heterogeneous devices, the priority here is pretty weak. Maybe we need scoring for beta? Even with scoring, it will be an *influence* but not a strict priority.

Reply:
I added some clarification. But I am also open to making scoring part of alpha.

Reply:
I think not having scoring for alpha is fine -- we will just need to be aware of this limitation in setups where it might matter. Where I actually see it being used the most though is to write ClaimTemplates that can be used in different environments with different types of GPUs without modification.

This is especially relevant for [NIMs](https://developer.nvidia.com/nim) which have different variants, optimized for specific hardware configurations (e.g. 1 H100, 2 A100s, 4 T4s, etc.). Depending on what hardware they have been given, they will pull / run with a different model optimized for that specific hardware configuration.

With this feature in place, our [NIM Operator](https://github.com/NVIDIA/k8s-nim-operator) can build a ResourceClaimTemplate with `FirstAvailableOf` requests for the GPU requirements of each optimized model of a given NIM. Then (depending on the available hardware) the scheduler will have the ability to place the NIM one the "best" node possible without having to put any custom logic for this in the operator itself.

Reply:
sgtm

/hold cancel